include/alp/utils/parser/MatrixFileIterator.hpp

```diff
@@ -0,0 +1,377 @@
+						throw std::runtime_error(
+							"Only sequential IO is supported by this iterator at "
+							"present, sorry."
+						);
+					}
+				}
+
+				/** Copy constructor. */
+				MatrixFileIterator( const MatrixFileIterator< T > & other ) :
+					properties( other.properties ), infile( properties._fn ), spos( other.spos ),
+					colidx( other.colidx ), rowidx( other.rowidx ), ended( other.ended ), started( other.started ),
+					converter( other.converter ) {
+					// set latest stream position
+					(void)infile.seekg( spos );
+				}
+
+				/** Base destructor. */
+				~MatrixFileIterator() {
+				}
+
+				/** Copies an iterator state. */
+				MatrixFileIterator & operator=( const MatrixFileIterator & x ) {
+					// copy ended state
+					ended = x.ended;
+					// copy started state
+					started = x.started;
+					// copy converter
+					converter = x.converter;
+					// check if we are done already
+					if( ended ) {
+						return *this;
+					}
+					// check if both iterators have a similar ifstream
+					if( properties._fn == x.properties._fn ) {
+						// NO, so re-create infile
+						if( infile.is_open() ) {
+							infile.close();
+						}
+						infile.open( x.properties._fn );
+						properties = x.properties;
+					}
+					// not yet done, copy input file stream position
+					(void)infile.seekg( x.spos );
+					spos = x.spos;
+					return *this;
+				}
+
+				/** Standard check for equality. */
+				bool operator==( const MatrixFileIterator &x ) const {
+					// check if both are in end position
+					if( ended && x.ended ) {
+						return true;
+					}
+#ifndef NDEBUG
+					if( properties._fn != x.properties._fn ) {
+						std::cerr << "Warning: comparing two instances of "
+									 "MatrixFileIterator that are 1) nonempty "
+									 "*and* 2) not reading from the same "
+									 "file.\n";
+					}
+#endif
+					// check if both are in start position
+					if( started && x.started ) {
+						return true;
+					}
+
+					// otherwise, not equal
+					return false;
+				}
+
+				/** Standard check for inequality, relies on equality check. */
+				bool operator!=( const MatrixFileIterator &x ) const {
+					return ! ( operator==( x ) );
+				}
+
+				// this assumes full triplet data
+				MatrixFileIterator & operator++() {
+					// if ended, stop
+					if( ended ) {
+						return *this;
+					}
+					// if this is the first function call on this iterator, call preprocess first
+					if( started ) {
+						preprocess();
+						started = false;
+						(void)operator++();
+					}
+
+					if( ! ( infile >> val ) ) {
+						ended = true;
+					} else {
+#ifdef _DEBUG
+						T temp = val;
+						converter( temp );
+						std::cout << "MatrixFileIterator::operator++  parsed line ``"
+							  << val  << "'', with value after conversion "
+							  << temp << "\n";
+#endif
+						// convert value
+						converter( val );
+
+						if(
+							properties._symmetry == MatrixFileProperties::MMsymmetries::SYMMETRIC ||
```

Comment:
Coding guidelines suggest to wrap `... == ...` in parenthesis: `( ... == ... )`.